### PR TITLE
[suspendplugin] Add support for devices which only need the wowlan suspending. JB#46941

### PR DESCRIPTION
**Pull request by mlehtima**

### CODE REVIEW

File: rpm/sailfish-connman-plugin-suspend-wmtwifi.spec [hunk 1]
```diff
@@ -17,22 +17,22 @@ BuildRequires: pkgconfig(libnl-genl-3.0)
...
-mkdir -p %{buildroot}/etc/udev/rules.d
-install -m 664 %{SOURCE1} %{buildroot}/etc/udev/rules.d/70-wifi-powersave.rules
+mkdir -p %{buildroot}/lib/udev/rules.d
+install -m 664 %{SOURCE1} %{buildroot}/lib/udev/rules.d/70-wifi-powersave.rules
 
 %files
 %defattr(-,root,root,-)
 %{plugin_dir}/*.so
-%{_sysconfdir}/udev/rules.d/70-wifi-powersave.rules
+/lib/udev/rules.d/70-wifi-powersave.rules
```
Comment (sledges) [inline rpm/sailfish-connman-plugin-suspend-wmtwifi.spec:39]: This and above related changes would look nicer in a separate commit

**Review by sledges — CHANGES_REQUESTED** [review]

File: src/suspendplugin.c [hunk 2]
```diff
@@ -70,7 +71,7 @@ struct testmode_cmd_suspend {
     struct testmode_cmd_hdr header;
     uint8_t suspend;
 };
-
+#endif
```
Comment (monich) [inline src/suspendplugin.c]: I would add an empty line after `#endif` for readability

**Comment by monich** [comment]
Should these two plugins exclude each other? Or could it ever make sense to have them both installed?

**Comment by monich** [comment]
Would it be feasible to choose the right behavior at run time?

**Comment by mlehtima** [comment]
Reworked to select used features at runtime.

File: rpm/sailfish-connman-plugin-suspend-wmtwifi.spec (hunk 1 above)
Comment (mlehtima) [reply rpm/sailfish-connman-plugin-suspend-wmtwifi.spec:39]: Changes moved to separate commit.

**Comment by monich** [comment]
Should we suspend when tethering is on? (esp. if charger is connected)

**Comment by sledges** [comment]
> Should we suspend when tethering is on? (esp. if charger is connected)

This reminded me one thing from my tests: Xperia 10 WLAN doesn't drain battery when Internet Sharing is on, only drains when connected to an AP.

**Comment by krnlyng** [comment]
> Should we suspend when tethering is on? (esp. if charger is connected)

on qcom we only set the wowlan parameters which afaict should always be set
on mtk we also set those suspend modes, if we only want those suspend modes to be set only when wifi is connected to an AP and not in sharing mode then we would check for the existence of the ap0 interface and set the modes based on that. the ap0 interface is dynamically added/removed when internet sharing is on/off on mtk afair

**Comment by monich** [comment]
> > Should we suspend when tethering is on? (esp. if charger is connected)
> 
> on qcom we only set the wowlan parameters which afaict should always be set
> on mtk we also set those suspend modes, if we only want those suspend modes to be set only when wifi is connected to an AP and not in sharing mode then we would check for the existence of the ap0 interface and set the modes based on that. the ap0 interface is dynamically added/removed when internet sharing is on/off on mtk afair

This ap0 thing is mtk specific afair, would it be better to somehow expose connman tethering status to plugins?

**Comment by krnlyng** [comment]
> > > Should we suspend when tethering is on? (esp. if charger is connected)
> > 
> > 
> > on qcom we only set the wowlan parameters which afaict should always be set
> > on mtk we also set those suspend modes, if we only want those suspend modes to be set only when wifi is connected to an AP and not in sharing mode then we would check for the existence of the ap0 interface and set the modes based on that. the ap0 interface is dynamically added/removed when internet sharing is on/off on mtk afair
> 
> This ap0 thing is mtk specific afair, would it be better to somehow expose connman tethering status to plugins?

yes the ap0 thing is mtk specific, but my point is that also the suspend calls are mtk specific, everything else, i.e. the wowlan, can (or should) always be set afaik

**Comment by monich** [comment]
> yes the ap0 thing is mtk specific, but my point is that also the suspend calls are mtk specific, everything else, i.e. the wowlan, can (or should) always be set afaik

true :)

**Comment by krnlyng** [comment]
When i look at the kernel sources i see that there are different functions for handling suspend in the kernel
when ap is active -> p2pHandleSystemSuspend is called
when normal mode is active -> wlanHandleSystemSuspend is called

So i think we are supposed to make these calls even in ap mode?

**Comment by monich** [comment]
> So i think we are supposed to make these calls even in ap mode?

Suspending in AP (tethering) mode would degrade the performance, wouldn't it?

**Comment by krnlyng** [comment]
> > So i think we are supposed to make these calls even in ap mode?
> 
> Suspending in AP (tethering) mode would degrade the performance, wouldn't it?

I assume so, but i am wondering why these functions exist in the first place.

**Review by krnlyng — APPROVED** [review]

**Review by sledges — APPROVED** [review]

**Pull request merged**